### PR TITLE
161114 application status refactor to use policy

### PR DESCRIPTION
**Pull request by weedySeaDragon**
PT Story:  **refactor 'Admin reviews application' to use Pundit policy for status** 
https://www.pivotaltracker.com/story/show/134291433

_This PR depends on PR #22  because it refactors/builds on it_ 

Changes proposed in this pull request:
1.  Use the MembershipApplicationPolicy to control who can edit the status
2. create a partial for editing the status (just pulled the existing form into a partial)
3. Simplified the status and timestamp (just downcase the status and add on the timestamp)

Ready for review
@thesuss & ??

### CODE REVIEW

File: app/views/membership_applications/show.html.haml [hunk 1]
```diff
@@ -8,3 +8,9 @@
   - if @membership_application.phone_number
     %p= @membership_application.phone_number
   %p= mail_to @membership_application.company_email, @membership_application.company_email
+  .wpcf7
+    Application Status:
+    - if policy(@membership_application).permitted_attributes_for_edit.include? :status
```
[the remark this answers is not in the document]
Comment (thesuss) [reply app/views/membership_applications/show.html.haml]: You can move this line before .wpcf
and put .wpcf in top of the new partial

**Review by thesuss — APPROVED** [review]
I think I understand how the policies work, thanks to the comments. Perhaps they should be removed before merge though?

File: app/controllers/membership_applications_controller.rb [hunk 2]
```diff
@@ -6,7 +6,10 @@ def new
   end
 
   def create
-    @membership_application = current_user.membership_applications.new(membership_application_params)
+    # must create the membership_application before using the policy to check the parameters. @see https://www.sitepoint.com/straightforward-rails-authorization-with-pundit/
+    @membership_application = current_user.membership_applications.new
```
Comment (tochman) [inline app/controllers/membership_applications_controller.rb]: Unless this code is WIP I would like to ask you to remove comments from the code.

File: app/controllers/membership_applications_controller.rb [hunk 3]
```diff
@@ -6,7 +6,10 @@ def new
...
+    @membership_application.update(permitted_attributes(@membership_application))
+
```
Comment (tochman) [inline app/controllers/membership_applications_controller.rb]: I don't think this is necessary. You could probably authorize the class directly by calling 

```
authorize MembershipApplication
```

I think I didi it on the MembershipApplicationController#index

File: app/policies/membership_application_policy.rb [hunk 4]
```diff
@@ -1,3 +1,48 @@
...
+  def permitted_attributes
+    if user && user.admin?
+      attributes_and_status
+    else
+      user_owner_attributes
+    end
+  end
+
+
+  # When a user creates a MembershipApplication, the status is initialized (so can create it)
```
Comment (tochman) [inline app/policies/membership_application_policy.rb]: Your tests are documenting you code. So comments are not necessary. ;-)

File: app/views/membership_applications/show.html.haml [hunk 5]
```diff
@@ -9,12 +9,8 @@
...
-                           ['Rejected', 'Rejected']])
-      = f.submit 'Update', class: 'wpcf7-submit'
-      - if @membership_application.status == 'Accepted'
-        %p Membership accepted at #{@membership_application.updated_at.strftime('%F')}
-      - if @membership_application.status == 'Rejected'
-        %p Membership rejected at #{@membership_application.updated_at.strftime('%F')}
\ No newline at end of file
+    Application Status:
+    - if policy(@membership_application).permitted_attributes_for_edit.include? :status
+      = render :partial => 'application_status_form'
```
Comment (tochman) [inline app/views/membership_applications/show.html.haml]: Old hash syntax

File: app/views/membership_applications/show.html.haml [hunk 6]
```diff
@@ -9,12 +9,8 @@
...
-                           ['Accepted', 'Accepted'],
-                           ['Rejected', 'Rejected']])
-      = f.submit 'Update', class: 'wpcf7-submit'
-      - if @membership_application.status == 'Accepted'
-        %p Membership accepted at #{@membership_application.updated_at.strftime('%F')}
-      - if @membership_application.status == 'Rejected'
-        %p Membership rejected at #{@membership_application.updated_at.strftime('%F')}
\ No newline at end of file
+    Application Status:
+    - if policy(@membership_application).permitted_attributes_for_edit.include? :status
```
Comment (tochman) [inline app/views/membership_applications/show.html.haml]: It feels like a bit too much logic for a view but IF  there's no other way to do this I'm okay with it. Extract this to a helper perhaps?

File: app/views/membership_applications/show.html.haml [hunk 7]
```diff
@@ -9,12 +9,8 @@
...
+      = render :partial => 'application_status_form'
+
+    %p Membership #{@membership_application.status.downcase} at #{@membership_application.updated_at.strftime('%F')}
```
Comment (tochman) [inline app/views/membership_applications/show.html.haml]: I would extract this to separate helpers.

File: spec/policies/membership_application_policy_spec.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+describe MembershipApplicationPolicy do
+
+  let(:application_owner) { create(:user, email: 'user_1@random.com') }
+  let(:admin) { create(:user, email: 'admin@sgf.com', admin: true) }
+
+  let(:application) { create(:membership_application,
+                             user: application_owner) }
+
+  subject { MembershipApplicationPolicy.new(application_owner, application) }
```
Comment (tochman) [inline spec/policies/membership_application_policy_spec.rb]: You are using `described_class` in other parts of the test. Why `MembershipApplicationPolicy` here?

**Review by tochman — CHANGES_REQUESTED** [review]

File: app/views/membership_applications/show.html.haml (hunk 5 above)
Comment (thesuss) [reply app/views/membership_applications/show.html.haml]: `render partial: 'application_status_form'`

File: spec/policies/membership_application_policy_spec.rb (hunk 8 above)
Comment (weedySeaDragon) [reply spec/policies/membership_application_policy_spec.rb]: just sloppiness :-)

File: app/views/membership_applications/show.html.haml (hunk 6 above)
Comment (weedySeaDragon) [reply app/views/membership_applications/show.html.haml]: agreed.  The `- if...` seems much too heavy since it invokes the policy.

**Comment by weedySeaDragon** [comment]
Fixed the items noted.

In finding a way to check the parameters during a create (= pass in the class to the policy instead of an instance), I had to do much more fine-grained testing.  (Which is a good thing!)  So the RSpec tests are much more fine-grained now. 

Ready for another review @thesuss  and @tochman (and whomever else)

File: spec/policies/membership_application_policy_spec.rb [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+    let(:application) { create(:membership_application,
+                               user: application_owner) }
+
+
+    describe 'For the MembershipApplication creator' do
+      it_behaves_like 'can see the status' do
+        let(:this_application) { application }
+        let(:this_user) { application_owner }
+      end
```
Comment (tochman) [inline spec/policies/membership_application_policy_spec.rb:56]: In this `it` block you are creating two objects but you have no assertion/expectations. What are you testing. (This goes for the it-blocks below as well)

**Review by tochman — CHANGES_REQUESTED** [review]

File: spec/policies/membership_application_policy_spec.rb (hunk 9 above)
Comment (thesuss) [reply spec/policies/membership_application_policy_spec.rb:56]: I understand it like they are referenced here. They really are above. Line 32?

Comment (weedySeaDragon) [reply spec/policies/membership_application_policy_spec.rb:56]: This uses the shared example `can see the status` (line 32) to test that the application owner is permitted to see the 'status' attribute.  Given the application, it tests that if the user is the application owner, the user (application owner) can see the _status_ parameter, according to the policy.

This it block (and others) uses these `let` statements to set the variables that are used in the shared example.  

Is that what you're asking?

Comment (weedySeaDragon) [reply spec/policies/membership_application_policy_spec.rb:56]: And if it's not clear (kind of aside from the question of _is this the right thing and is it correct_), is there something I can do to make it more clear?

Comment (tochman) [reply spec/policies/membership_application_policy_spec.rb:56]: Okay, I think that this introduces too much complexity to these specs but I can follow the logic. Using shared examples is an interesting approach that I would like to make sure that everybody is onboard with before  we start introducing. I totally see the need to keep specs DRY, but I think it's a bit to early. I will not consider it a blocker though.

**Review by tochman — APPROVED** [review]

File: spec/policies/membership_application_policy_spec.rb (hunk 9 above)
Comment (weedySeaDragon) [reply spec/policies/membership_application_policy_spec.rb:56]: I have no problem changing these to not use shared_examples.  I think that RSpec's shared_examples can be unclear; they make the narrative (wording) awkward.  
This is a good example:  this didn't read smoothly and didn't come across clearly to you.

We can merge this so things keep moving forward.  I'll add a chore in PT to refactor (not use shared_examples) and reference this discussion.

**Pull request merged**